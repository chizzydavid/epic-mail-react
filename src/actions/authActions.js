import { 
  SET_CURRENT_USER, 
  AUTH_ERROR, 
  INIT_AUTH_REQUEST 
} from './types';
import axios from 'axios';
import setAuthToken from '../utils/setAuthToken';

const authUrl = `https://chizzy-epicmail.herokuapp.com/api/v2/auth/`;

export const signUpUser = (payload) => dispatch => {
  dispatch({type: INIT_AUTH_REQUEST}); 

  axios.post(`${authUrl}signup`, payload)
    .then((res) => {
      const { token, user } = res.data.data;
      localStorage.setItem('token', token);

      setAuthToken(token);
      dispatch(setCurrentUser(user));
    })
    .catch(err => {
      console.log(err);
      dispatch({
        type: AUTH_ERROR,
        payload: err.response.data
      });
    });
}

export const setCurrentUser = payload => {
  return {
    type: SET_CURRENT_USER,
    payload: payload
  }
}
