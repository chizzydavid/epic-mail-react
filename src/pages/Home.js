import React from 'react';
import Navbar from '../components/shared/Navbar';
import Landing from '../components/Landing';
import Footer from '../components/shared/Footer';

const Home = () => {
  return (
    <div id="home">
      <Navbar home={true}/>
      <Landing/>
      <Footer />
    </div>
  )
}

export default Home;