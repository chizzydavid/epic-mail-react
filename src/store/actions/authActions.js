import { 
  SET_CURRENT_USER, 
  AUTH_ERROR, 
  INIT_AUTH_REQUEST 
} from './types';
import axios from 'axios';
import setAuthToken from '../../utils/setAuthToken';

const authUrl = `https://chizzy-epicmail.herokuapp.com/api/v2/auth/`;

export const signUpUser = (payload) => dispatch => {
  dispatch({type: INIT_AUTH_REQUEST}); 

  return axios.post(`${authUrl}signup`, payload)
    .then((res) => {
      const { token, user } = res.data.data;
      localStorage.setItem('token', token);

      setAuthToken(token);
      dispatch(setCurrentUser(user));
    })
    .catch(err => {
      dispatch({
        type: AUTH_ERROR,
        payload: err.response.data
      });
    });
}

export const setCurrentUser = payload => {
  return {
    type: SET_CURRENT_USER,
    payload: payload
  }
}

export const loginUser = (payload) => dispatch => {
  dispatch({type: INIT_AUTH_REQUEST});

  return axios.post(`${authUrl}login`, payload)
    .then((res) => {
      const { token, user } = res.data.data;
      localStorage.setItem('token', token);

      setAuthToken(token);
      dispatch(setCurrentUser(user));
    })
    .catch(err => {
      dispatch({
        type: AUTH_ERROR,
        payload: err.response.data
      })
    }) 
}

