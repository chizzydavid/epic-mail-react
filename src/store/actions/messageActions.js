import {
  BEGIN_MSG_REQUEST,
  MSG_ERROR,
  MSG_FEEDBACK,
  SEND_MSG_FEEDBACK,
  SET_RECEIVED_MESSAGES,
  SET_UNREAD_MESSAGES,
  SET_READ_MESSAGES,
  SET_SENT_MESSAGES,
  SET_DRAFTS,
  SET_ACTIVE_TAB,
  REMOVE_MESSAGE_ALL
} from './types';
import { BASE_URL } from '../constants';
import axios from 'axios';

export const getMessages = (category) => dispatch => {
  dispatch({type: BEGIN_MSG_REQUEST}); 
  const msgCategory = category === 'all' ? '' : category;

  return axios.get(`${BASE_URL}messages/${msgCategory}`)
    .then(res => {
      dispatch(setActiveMsgTab(category));
      if (res.data.message) {
        dispatch(messageFeedback(res.data.message));
        return;
      }
      dispatch(setMessages(category, res.data.data));
    })
    .catch(err => {
      dispatch(messageError(err.response.data));
    });
}

export const sendMessage = (payload) => dispatch => {
  dispatch({type: BEGIN_MSG_REQUEST}); 

  axios.post(`${BASE_URL}messages/`, payload)
    .then((res) => {
      if (res.data.status === 201) {
        dispatch(sendMessageFeedback('Message sent successfully.'))
      }
      dispatch(setMessages('sent', res.data));
    })
    .catch(err => {
      dispatch(messageError(err.response.data))
    });
}

export const deleteMessage = (messageId) => dispatch => {
  dispatch({type: BEGIN_MSG_REQUEST}); 

  axios.delete(`${BASE_URL}messages/${messageId}`)
    .then((res) => {
      if (res.data.status === 200) {
        dispatch({
          type: REMOVE_MESSAGE_ALL,
          payload: messageId,
        });
      }
    })
    .catch(err => {
      dispatch(messageError(err.response.data))
    });
}

export const setActiveMsgTab = category  => {
  return {
    type: SET_ACTIVE_TAB,
    payload: category
  }
}

export const setMessages = (category, payload) => {
  let type = category === 'all' ? SET_RECEIVED_MESSAGES : 
    category === 'unread' ? SET_UNREAD_MESSAGES : 
    category === 'read' ? SET_READ_MESSAGES : 
    category === 'sent' ? SET_SENT_MESSAGES : 
    category === 'drafts' ? SET_DRAFTS : '';

  return {
    type: type,
    payload: payload
  }
}

export const messageError = payload => {
  return {
    type: MSG_ERROR,
    payload: payload
  }
}

export const messageFeedback = payload => {
  return {
    type: MSG_FEEDBACK,
    payload: payload
  }
}

export const sendMessageFeedback = payload => {
  return {
    type: SEND_MSG_FEEDBACK,
    payload: payload
  }
}
