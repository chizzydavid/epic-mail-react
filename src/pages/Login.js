import React, { Component } from 'react'
import Navbar from '../components/shared/Navbar';
import Footer from '../components/shared/Footer';
import TextInput from '../components/shared/TextInput';
import Button from '../components/shared/Button';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link, Redirect } from 'react-router-dom';
import { loginUser } from '../store/actions/authActions';

export class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      validationErrors: {}
    };

    this.handleChange = this.handleChange.bind(this);
    this.validateFormInput = this.validateFormInput.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e) {
    const { validationErrors } = this.state;
    this.setState({
      [e.target.name]: e.target.value,
      validationErrors: {
        ...validationErrors,
        [e.target.name]: '',
      },      
    })
  }

  validateFormInput() {
    let errors = {};
    const emailRegx = /^\S+@\S+\.[\w]+$/;

    Object.entries(this.state).forEach(([property, value]) => {
      const input = typeof value === 'string' ? value.trim() : '';

      switch(property) {
        case 'email':
          if (input.length === 0) {
            errors[property] = 'Email address is required'
          }
          else if (!emailRegx.test(input)) {
              errors[property] = 'Enter a valid email address'
          }
          break;          
        case 'password':
          if (input.length === 0) {
            errors[property] = 'Password is required'
          }
          break;
        default:
          break;     
      }
    });

    if (Object.keys(errors).length !== 0) {
      this.setState({
        validationErrors: {...errors}
      })
      return false;
    } else {
      this.setState({
        validationErrors: {}
      })
      return true;
    }
  }

  handleSubmit(e) {
    e.preventDefault();

    if (this.validateFormInput()) {

      const newUser = Object.entries(this.state).reduce(
        (user, [property, input]) => {
          if (typeof input === 'string') {
            user[property] = input.trim();
          }
          return user;
        }, {})
      
      this.props.loginUser(newUser);
    }
  }

  render() {
    const { isLoading, isAuthenticated, errors } = this.props.auth;
    const { email, password } = this.state.validationErrors;

    if (isAuthenticated) {
      return <Redirect to="/dashboard" />;
    }        
    const authError = errors.error || '';

    return (
      <React.Fragment>
        <Navbar />
        <div className="container">    
          <form onSubmit={this.handleSubmit} id="form">
            <div>
              <h3> Login </h3>
              <span className="error" id="form-feedback">{authError}</span>

              <TextInput 
                error={email ? email : ''} 
                handleChange={this.handleChange} 
                type="text" 
                name="email" 
                placeholder="Enter email"
                label="Email" 
              />
              <TextInput 
                error={password ? password : ''} 
                handleChange={this.handleChange} 
                type="password" 
                name="password" 
                placeholder="Enter password"
                label="Password"
              />

              <Button
                type="submit"
                classname="submit"
                text={isLoading ? 'Please wait..' : 'Submit'}
              />              
              <p><Link className="sign-link forgot-password" to="/login">Forgot Password?</Link></p>  
              <p>Dont have an account yet? <Link className="sign-link" to="/sign-up">Sign Up</Link></p>
            </div>
          </form>
        </div>
        <Footer />
      </React.Fragment>
    )
  }
}

Login.propTypes = {
  loginUser: PropTypes.func.isRequired,
  auth: PropTypes.object
}

const mapStateToProps = (state) => ({
  auth: state.auth,
});

const mapDispatchToProps = {
  loginUser
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Login);
