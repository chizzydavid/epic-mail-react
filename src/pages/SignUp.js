import React, { Component } from 'react'
import Navbar from '../components/shared/Navbar';
import Footer from '../components/shared/Footer';
import TextInput from '../components/shared/TextInput';
import Button from '../components/shared/Button';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link, withRouter, Redirect } from 'react-router-dom';
import { signUpUser } from '../store/actions/authActions';

export class SignUp extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      confirmPassword: '',
      errors: {}
    };

    this.handleChange = this.handleChange.bind(this);
    this.validateFormInput = this.validateFormInput.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e) {
    const { errors } = this.state;

    this.setState({
      [e.target.name]: e.target.value,
      errors: {
        ...errors, 
        [e.target.name]: '',
      },
    })
  }

  validateFormInput() {
    let errors = {};
    const nameRegx = /^[a-zA-Z]{2,}$/;
    const emailRegx = /^\S+@\S+\.[\w]+$/;
    const passwordRegx = /^[\w]{6,20}$/;

    Object.entries(this.state).forEach(([property, value]) => {
      const input = typeof value === 'string' ? value.trim() : '';

      switch(property) {
        case 'firstName':
          if (input.length === 0) {
            errors[property] = 'First name is required'
          }
          else if(!nameRegx.test(input)) {
            errors[property] = 'First name must be at least 2 characters in length and contain only letters.'
          }
          break;
        case 'lastName':
          if (input.length === 0) {
            errors[property] = 'Last name is required.'
          }
          else if(!nameRegx.test(input)) {
            errors[property] = 'Last name must be at least 2 characters in length and contain only letters.'
          }          
          break;
        case 'email':
          if (input.length === 0) {
            errors[property] = 'Email address is required'
          }
          else if (!emailRegx.test(input)) {
              errors[property] = 'Enter a valid email address'
          }
          break;          
        case 'password':
          if (input.length === 0) {
            errors[property] = 'Password is required'
          }          
        else if (!passwordRegx.test(input)) {
            errors[property] = 'Password must be at least 6 characters in length and contain only letters and numbers '
          }
          break;
        case 'confirmPassword':
          if (this.state.password !== input) {
            errors[property] = 'Your two passwords dont match.'
          }
          break;
        default:
          break;     
      }
    });

    if (Object.keys(errors).length !== 0) {
      this.setState({
        errors: {...errors}
      })
      return false;
    } else {
      this.setState({ errors: {} })
      return true;
    }
  }

  handleSubmit(e) {
    e.preventDefault();

    if (this.validateFormInput()) {
      const newUser = Object.entries(this.state).reduce(
        (user, [property, input]) => {
          if (typeof input === 'string') {
            user[property] = input.trim();
          }
          return user;
        }, {})

      this.props.signUpUser(newUser);
    }
  }

  render() {
    const { isLoading, isAuthenticated, errors } = this.props.auth;
    if (isAuthenticated) {
      return <Redirect to="/dashboard" />;
    }    
    const { 
      firstName, lastName, email, password, confirmPassword 
    } = this.state.errors;
    
    const authError = errors.error || '';

    return (
      <React.Fragment>
        <Navbar />
        <div className="container">    
          <form onSubmit={this.handleSubmit} id="form" encType="multipart/form-data">
            <div>
              <h3> Create Account</h3>
              <span className="error" id="form-feedback">{authError}</span>

              <TextInput 
                error={firstName ? firstName : ''} 
                handleChange={this.handleChange} 
                type="text" 
                name="firstName" 
                placeholder="First name" 
                label="First Name" 
              />

              <TextInput 
                error={lastName ? lastName : ''} 
                handleChange={this.handleChange} 
                type="text" 
                name="lastName" 
                placeholder="Last name" 
                label="Last Name" 
              />

              <TextInput 
                error={email ? email : ''} 
                handleChange={this.handleChange} 
                type="text" 
                name="email" 
                placeholder="Enter email" 
                label="Email" 
              />

              <TextInput 
                error={password ? password : ''} 
                handleChange={this.handleChange} 
                type="password" 
                name="password" 
                placeholder="Enter password" 
                label="Password"
              />

              <TextInput 
                error={confirmPassword ? confirmPassword : ''} 
                handleChange={this.handleChange} 
                type="password" 
                name="confirmPassword" 
                placeholder="Confirm password" 
                label="Confirm Password" 
              />


              <p className="label">Upload Image</p>
              <img  width="150px" id="image-preview" />        
              <input id="image-upload" name="photo" type="file" accept="image/*" />

              <Button
                type="submit"
                classname="submit"
                text={isLoading ? 'Please wait..' : 'Submit'}
              />
              <p>Already have an account? <Link className="sign-link" to="/login">Log In</Link></p>
            </div>
          </form>
        </div>
        <Footer />
      </React.Fragment>
    )
  }
}

SignUp.propTypes = {
  signUpUser: PropTypes.func.isRequired,
  auth: PropTypes.object
}

const mapStateToProps = (state) => ({
  auth: state.auth,
});

const mapDispatchToProps = {
  signUpUser
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(SignUp));
