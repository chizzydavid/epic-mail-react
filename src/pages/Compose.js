import React, { Component } from 'react'
import Navbar from '../components/shared/Navbar';
import Button from '../components/shared/Button';
import Footer from '../components/shared/Footer';
import TextInput from '../components/shared/TextInput';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { sendMessage } from '../store/actions/messageActions';

export class Compose extends Component {
  constructor(props) {
    super(props);
    this.state = {
      recipient: '',
      subject: '',
      message: '',
      validationErrors: {},
    }
    this.handleChange = this.handleChange.bind(this);
    this.validateFormInput = this.validateFormInput.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e) {
    const { validationErrors } = this.state;

    this.setState({
      [e.target.name]: e.target.value,
      validationErrors: {
        ...validationErrors,
        [e.target.name]: '',
      },
    })
  }

  validateFormInput() {
    let errors = {};
    const emailRegx = /^\S+@\S+\.[\w]+$/;

    Object.entries(this.state).forEach(([property, value]) => {
      const input = typeof value === 'string' ? value.trim() : '';

      switch(property) {
        case 'recipient':
          if (input.length === 0) {
            errors[property] = 'Message recipient is required'
          }
          else if(!emailRegx.test(input)) {
            errors[property] = 'Invalid email address for recipient'
          }
          break;
        case 'subject':
          if (input.length === 0) {
            errors[property] = 'Subject is required'
          }
          break;
        case 'message':
          if (input.length === 0) {
            errors[property] = 'Enter a message body'
          }
          break;          
        default:
          break;     
      }
    });

    if (Object.keys(errors).length !== 0) {
      this.setState({
        validationErrors: {...errors}
      })
      return false;
    } else {
      this.setState({
        validationErrors: {}
      })
      return true;
    }
  }

  handleSubmit(e) {
    e.preventDefault();

    if (this.validateFormInput()) {

      const newMessage = Object.entries(this.state).reduce(
        (user, [property, input]) => {
          if (typeof input === 'string') {
            user[property] = input.trim();
          }
          return user;
        }, {})

      this.props.sendMessage(newMessage);
    }
  }

  render() {
    const { recipient, subject, message } = this.state.validationErrors;
    const { isLoading, sendMsgFeedback, errors } = this.props.messages;
    const [formFeedback, feedbackClass] = sendMsgFeedback ? [sendMsgFeedback, 'success'] : 
      errors.error ? [errors.error, 'error'] : '';

    return (
      <React.Fragment>
        <Navbar />
        <div className="compose-container">
          <p className="close-icon"> 
            <i
              onClick={this.props.close} 
              className="fa fa-close"
            >
            </i>
          </p>
          <div className="c-container">
            <form onSubmit={this.handleSubmit} id="send-message-form">
              <h3 className="">Create New Message</h3>
              <div>
                <p className={feedbackClass} id="form-feedback">{formFeedback}</p>

                <TextInput 
                  error={recipient ? recipient : ''} 
                  handleChange={this.handleChange} 
                  type="text" 
                  name="recipient"
                  placeholder="Enter Recipient"
                  label="To" 
                />

                <TextInput 
                  error={subject ? subject : ''} 
                  handleChange={this.handleChange} 
                  type="text" 
                  name="subject" 
                  placeholder="Subject" 
                  label="Subject" 
                />

                <p className="label">Message <i className="fa fa-asterisk"></i></p>
                <textarea onChange={this.handleChange} name="message" placeholder="Message" rows="8" cols="30"></textarea>
                <span className="error">{message ? message : ''}</span>
                
                <div id="compose-btn-wrapper"> 
                  <Button
                    id="send-message"
                    type="submit"
                    classname="submit"
                    text={isLoading ? 'Please wait..' : 'Send Message'}
                  />
                </div>
              </div>
            </form>
          </div>
        
        </div>
        <Footer />
      </React.Fragment>
    )
  }
}

Compose.propTypes = {
  sendMessage: PropTypes.func.isRequired,
  messages: PropTypes.object.isRequired,
  close: PropTypes.func
}

const mapStateToProps = (state) => ({
  messages: state.messages,
});

const mapDispatchToProps = {
  sendMessage
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Compose);
