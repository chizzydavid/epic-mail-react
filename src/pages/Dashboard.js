import React, { Component } from 'react'
import Navbar from '../components/shared/Navbar';
import Footer from '../components/shared/Footer';
import MessagesContainer from '../components/MessagesContainer';
import Compose from './Compose';

class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isShowing: false
    }

    this.openMsgCompose = this.openMsgCompose.bind(this);
    this.closeMsgCompose = this.closeMsgCompose.bind(this);
  }

  openMsgCompose() {
    this.setState({
      isShowing: true,
    })
  }

  closeMsgCompose() {
    this.setState({
      isShowing: false,
    })
  }

  render() {
    return (
      <React.Fragment>
        <Navbar
          open={this.openMsgCompose}
        />
        {this.state.isShowing && 
          <Compose
            close={this.closeMsgCompose}
          />
        }
        <MessagesContainer 
          open={this.openMsgCompose}        
        />
        <Footer />
      </React.Fragment>
    )
  }
}

export default Dashboard;
