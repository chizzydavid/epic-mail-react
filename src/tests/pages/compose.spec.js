import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import { Compose } from '../../pages/Compose';
import { sendMessage } from '../../store/actions/messageActions';

function shallowSetup() {
  const props = {
    messages: {
      isLoading: false,
      activeMsgTab: 'all',
      allReceived: [],
      unread: [],
      read: [],
      sent: [],
      feedback: '',
      sendMsgFeedback: '',
      errors: {
        error: 'Message recipient does not exist',
      },
    },
    sendMessage,
    close: () => {}
  };
  const enzymeWrapper = shallow(<Compose {...props} />);

  return {
    props,
    enzymeWrapper,
  };
}

describe('Compose component', () => {
  it('should render compose correctly', () => {
    const { enzymeWrapper, props: { messages: {errors: {error} } } } = shallowSetup();
    expect(enzymeWrapper).toMatchSnapshot();
    expect(enzymeWrapper.find('#form-feedback').text()).toBe(error);
  });
});

describe('Compose component methods', () => {
  beforeEach(() => {
    sinon.spy(Compose.prototype, 'handleChange');
    sinon.spy(Compose.prototype, 'handleSubmit');
  });

  afterEach(() => {
    Compose.prototype.handleChange.restore();
    Compose.prototype.handleSubmit.restore();
  });

  it('should call handlechange ', () => {
    const { enzymeWrapper } = shallowSetup();

    enzymeWrapper.find('TextInput').first().dive().find('input')
    .simulate('change', {
      target: {
        value: 'davidchizindu@gmail.com',
        name: 'recipient',
      },
    });
    expect(enzymeWrapper.state().recipient).toEqual('davidchizindu@gmail.com');
    expect(Compose.prototype.handleChange.calledOnce).toBe(true);
  });

  it('should call handlesubmit', () => {
    const { enzymeWrapper } = shallowSetup();
    enzymeWrapper.setState({
      recipient: 'davidchizindu@gmail.com',
      subject: 'How You Man',
      message: 'How is everything going',
    });

    const form = enzymeWrapper.find('#send-message-form');
    const mockedEvent = {
      target: {},
      preventDefault: () => ({}),
    };

    form.simulate('submit', mockedEvent);
    expect(Compose.prototype.handleSubmit.calledOnce).toBe(true);
  });

  it('should handle validation after form is submitted', () => {
    const { enzymeWrapper } = shallowSetup();
    enzymeWrapper.setState({
      recipient: '',
      subject: '',
      message: '',
    });

    const form = enzymeWrapper.find('#send-message-form');
    const mockedEvent = {
      target: {},
      preventDefault: () => ({}),
    };

    form.simulate('submit', mockedEvent);
    expect(Compose.prototype.handleSubmit.calledOnce).toBe(true);
  });
});
