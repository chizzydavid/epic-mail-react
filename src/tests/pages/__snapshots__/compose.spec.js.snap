// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compose component should render compose correctly 1`] = `ShallowWrapper {}`;
